opengovsg/postmangovsg
# Log email bounce and complaint subtype

## Problem

It would be useful to log the bounce/complaint subtype for email bounces to understand the reason behind the bounce. This also allows us to know if a blacklisted email is part of SES's global suppression list, in the case we need to remove a blacklisted email.

Closes #1033 

## Solution

**Improvements**:

- Add a new column in `email_messages` to log error sub types
- Log the email bounce subtype and complain subtype

## Tests

- Ensure the following recipients do not exist in the `email_blacklist` table
- Send an email campaign to the recipients
- Check the db for the `error_sub_type` of the email messages sent

```
recipient

// SES
bounce@simulator.amazonses.com
complaint@simulator.amazonses.com
suppressionlist@simulator.amazonses.com

// SendGrid
bouncetest@tribulant.com
bounce-test@service.socketlabs.com
```

Documentation on these mailbox simulator addresses can be found [here](https://docs.aws.amazon.com/ses/latest/DeveloperGuide/send-email-simulator.html)

**Note**
I expected a different error_sub_type for `suppressionlist@simulator.amazonses.com` but it was classified as `General` too

## Deploy Notes

- Create a new column `error_sub_type` of type `varchar(255)` and is nullable to the `email_messages` and `email_ops` tables

**Migration query:**
```sql
ALTER TABLE email_messages
  ADD error_sub_type VARCHAR(255);

ALTER TABLE email_ops
  ADD error_sub_type VARCHAR(255);
```
